shared/src/balancer/event_handler.rs

```diff
@@ -101,27 +107,18 @@ impl PoolDataFetching for Web3 {
         batch.execute_all(MAX_BATCH_SIZE).await;
 
         let tokens = token_data.await?.0;
-        // This is already a batch call for a list of token decimals.
-        let mut batch = CallBatch::new(self.transport());
-        let futures = tokens
+
+        let token_decimals = self.token_info_fetcher.get_token_infos(&tokens).await;
+        // Note that balancer does not support tokens with more than 18 decimals
+        // https://github.com/balancer-labs/balancer-v2-monorepo/blob/ce70f7663e0ac94b25ed60cb86faaa8199fd9e13/pkg/pool-utils/contracts/BasePool.sol#L497-L508
+        let ordered_decimals = tokens
             .iter()
-            .map(|address| {
-                let erc20 = ERC20::at(&self, *address);
-                erc20.methods().decimals().batch_call(&mut batch)
-            })
-            .collect::<Vec<_>>();
-        batch.execute_all(MAX_BATCH_SIZE).await;
-        let token_decimals = join_all(futures)
-            .await
-            .into_iter()
-            .collect::<Result<Vec<_>, _>>()?;
-        let scaling_exponents = token_decimals
+            .map(|token| token_decimals.get(token).and_then(|t| t.decimals))
+            .collect::<Option<Vec<_>>>()
+            .ok_or_else(|| anyhow!("all token decimals required to build scaling factors"))?;
+        let scaling_exponents = ordered_decimals
             .iter()
-            .map(|decimals| {
-                // Note that balancer does not support tokens with more than 18 decimals
-                // https://github.com/balancer-labs/balancer-v2-monorepo/blob/ce70f7663e0ac94b25ed60cb86faaa8199fd9e13/pkg/pool-utils/contracts/BasePool.sol#L497-L508
-                18 - decimals
```

Comment:
I think the comment is still relative.
Also, I relized that this is slightly incorrect and should be:
```
18.checked_sub(decimals).ok_or_else(|| anyhow!("token with more than 18 decimals"))
```

Otherwise you could get wrapping behaviour in release mode.